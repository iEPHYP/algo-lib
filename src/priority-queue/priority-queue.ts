import { MinHeap } from '@/heap';

export interface PriorityQueueItem {
  priority: number;
}

// TODO: items with the same priority don't get dequeued properly
// TODO: extract comparing function to jungle priorities according to any item props
export class PriorityQueue<TItem extends PriorityQueueItem = PriorityQueueItem> {
  private heap = new MinHeap<TItem>({ getValue: (item) => item.priority });

  // Complexity O(logN)
  public enqueue(item: TItem) {
    this.heap.insert(item);
  }

  // Complexity O(logN)
  public dequeue() {
    return this.heap.extractMin() || null;
  }

  public toList() {
    return this.heap.toList();
  }

  public size() {
    return this.heap.size();
  }
}
