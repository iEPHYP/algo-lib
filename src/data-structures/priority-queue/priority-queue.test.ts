import { PriorityQueue, PriorityQueueItem } from './priority-queue';

interface Patient extends PriorityQueueItem {
  name: string;
  diagnosis: string;
}

describe('PriorityQueue', () => {
  let hospitalPriorityQueue = new PriorityQueue<Patient>();

  const fillPriorityQueue = () => {
    hospitalPriorityQueue.enqueue({ name: 'Simon', diagnosis: 'injury', priority: 3 });
    hospitalPriorityQueue.enqueue({ name: 'Eric', diagnosis: 'exploded head', priority: 0 });
    hospitalPriorityQueue.enqueue({ name: 'Aaron', diagnosis: 'sniffing', priority: 6 });
    hospitalPriorityQueue.enqueue({ name: 'Jane', diagnosis: 'bleeding', priority: 2 });
    hospitalPriorityQueue.enqueue({ name: 'Nick', diagnosis: 'flu', priority: 4 });
    hospitalPriorityQueue.enqueue({ name: 'May', diagnosis: 'flu', priority: 4 });
    hospitalPriorityQueue.enqueue({ name: 'Jack', diagnosis: 'high temperature', priority: 8 });
    hospitalPriorityQueue.enqueue({ name: 'Dread', diagnosis: 'flu', priority: 4 });
  };

  beforeEach(() => {
    hospitalPriorityQueue = new PriorityQueue<Patient>();
  });

  it('enqueues properly', () => {
    fillPriorityQueue();

    expect(hospitalPriorityQueue.toList()).toMatchSnapshot();
  });

  it('dequeues properly', () => {
    expect(hospitalPriorityQueue.dequeue()).toMatchSnapshot('value on empty queue');
    expect(hospitalPriorityQueue.toList()).toMatchSnapshot('after empty queue');
    fillPriorityQueue();
    expect(hospitalPriorityQueue.dequeue()).toMatchSnapshot('value on full queue');
    expect(hospitalPriorityQueue.toList()).toMatchSnapshot('on full queue');
    expect(hospitalPriorityQueue.dequeue()).toMatchSnapshot('value on full queue');
    expect(hospitalPriorityQueue.toList()).toMatchSnapshot('on full queue');
    expect(hospitalPriorityQueue.dequeue()).toMatchSnapshot('value on full queue');
    expect(hospitalPriorityQueue.toList()).toMatchSnapshot('on full queue');
    expect(hospitalPriorityQueue.dequeue()).toMatchSnapshot('value on full queue');
    expect(hospitalPriorityQueue.toList()).toMatchSnapshot('on full queue');
    expect(hospitalPriorityQueue.dequeue()).toMatchSnapshot('value on full queue');
    expect(hospitalPriorityQueue.toList()).toMatchSnapshot('on full queue');
    expect(hospitalPriorityQueue.dequeue()).toMatchSnapshot('value on full queue');
    expect(hospitalPriorityQueue.toList()).toMatchSnapshot('on full queue');
    expect(hospitalPriorityQueue.dequeue()).toMatchSnapshot('value on full queue');
    expect(hospitalPriorityQueue.toList()).toMatchSnapshot('on full queue');
    expect(hospitalPriorityQueue.dequeue()).toMatchSnapshot('value on full queue');
    expect(hospitalPriorityQueue.toList()).toMatchSnapshot('on full queue');
    expect(hospitalPriorityQueue.dequeue()).toMatchSnapshot('value on full queue');
    expect(hospitalPriorityQueue.toList()).toMatchSnapshot('on full queue');
  });
});
