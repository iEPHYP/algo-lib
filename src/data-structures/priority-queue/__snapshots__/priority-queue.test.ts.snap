// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriorityQueue dequeues properly: after empty queue 1`] = `[]`;

exports[`PriorityQueue dequeues properly: on full queue 1`] = `
[
  {
    "diagnosis": "bleeding",
    "name": "Jane",
    "priority": 2,
  },
  {
    "diagnosis": "injury",
    "name": "Simon",
    "priority": 3,
  },
  {
    "diagnosis": "flu",
    "name": "May",
    "priority": 4,
  },
  {
    "diagnosis": "flu",
    "name": "Dread",
    "priority": 4,
  },
  {
    "diagnosis": "flu",
    "name": "Nick",
    "priority": 4,
  },
  {
    "diagnosis": "sniffing",
    "name": "Aaron",
    "priority": 6,
  },
  {
    "diagnosis": "high temperature",
    "name": "Jack",
    "priority": 8,
  },
]
`;

exports[`PriorityQueue dequeues properly: on full queue 2`] = `
[
  {
    "diagnosis": "injury",
    "name": "Simon",
    "priority": 3,
  },
  {
    "diagnosis": "flu",
    "name": "Nick",
    "priority": 4,
  },
  {
    "diagnosis": "flu",
    "name": "May",
    "priority": 4,
  },
  {
    "diagnosis": "flu",
    "name": "Dread",
    "priority": 4,
  },
  {
    "diagnosis": "high temperature",
    "name": "Jack",
    "priority": 8,
  },
  {
    "diagnosis": "sniffing",
    "name": "Aaron",
    "priority": 6,
  },
]
`;

exports[`PriorityQueue dequeues properly: on full queue 3`] = `
[
  {
    "diagnosis": "flu",
    "name": "May",
    "priority": 4,
  },
  {
    "diagnosis": "flu",
    "name": "Nick",
    "priority": 4,
  },
  {
    "diagnosis": "sniffing",
    "name": "Aaron",
    "priority": 6,
  },
  {
    "diagnosis": "flu",
    "name": "Dread",
    "priority": 4,
  },
  {
    "diagnosis": "high temperature",
    "name": "Jack",
    "priority": 8,
  },
]
`;

exports[`PriorityQueue dequeues properly: on full queue 4`] = `
[
  {
    "diagnosis": "flu",
    "name": "Nick",
    "priority": 4,
  },
  {
    "diagnosis": "flu",
    "name": "Dread",
    "priority": 4,
  },
  {
    "diagnosis": "sniffing",
    "name": "Aaron",
    "priority": 6,
  },
  {
    "diagnosis": "high temperature",
    "name": "Jack",
    "priority": 8,
  },
]
`;

exports[`PriorityQueue dequeues properly: on full queue 5`] = `
[
  {
    "diagnosis": "flu",
    "name": "Dread",
    "priority": 4,
  },
  {
    "diagnosis": "high temperature",
    "name": "Jack",
    "priority": 8,
  },
  {
    "diagnosis": "sniffing",
    "name": "Aaron",
    "priority": 6,
  },
]
`;

exports[`PriorityQueue dequeues properly: on full queue 6`] = `
[
  {
    "diagnosis": "sniffing",
    "name": "Aaron",
    "priority": 6,
  },
  {
    "diagnosis": "high temperature",
    "name": "Jack",
    "priority": 8,
  },
]
`;

exports[`PriorityQueue dequeues properly: on full queue 7`] = `
[
  {
    "diagnosis": "high temperature",
    "name": "Jack",
    "priority": 8,
  },
]
`;

exports[`PriorityQueue dequeues properly: on full queue 8`] = `[]`;

exports[`PriorityQueue dequeues properly: on full queue 9`] = `[]`;

exports[`PriorityQueue dequeues properly: value on empty queue 1`] = `null`;

exports[`PriorityQueue dequeues properly: value on full queue 1`] = `
{
  "diagnosis": "exploded head",
  "name": "Eric",
  "priority": 0,
}
`;

exports[`PriorityQueue dequeues properly: value on full queue 2`] = `
{
  "diagnosis": "bleeding",
  "name": "Jane",
  "priority": 2,
}
`;

exports[`PriorityQueue dequeues properly: value on full queue 3`] = `
{
  "diagnosis": "injury",
  "name": "Simon",
  "priority": 3,
}
`;

exports[`PriorityQueue dequeues properly: value on full queue 4`] = `
{
  "diagnosis": "flu",
  "name": "May",
  "priority": 4,
}
`;

exports[`PriorityQueue dequeues properly: value on full queue 5`] = `
{
  "diagnosis": "flu",
  "name": "Nick",
  "priority": 4,
}
`;

exports[`PriorityQueue dequeues properly: value on full queue 6`] = `
{
  "diagnosis": "flu",
  "name": "Dread",
  "priority": 4,
}
`;

exports[`PriorityQueue dequeues properly: value on full queue 7`] = `
{
  "diagnosis": "sniffing",
  "name": "Aaron",
  "priority": 6,
}
`;

exports[`PriorityQueue dequeues properly: value on full queue 8`] = `
{
  "diagnosis": "high temperature",
  "name": "Jack",
  "priority": 8,
}
`;

exports[`PriorityQueue dequeues properly: value on full queue 9`] = `null`;

exports[`PriorityQueue enqueues properly 1`] = `
[
  {
    "diagnosis": "exploded head",
    "name": "Eric",
    "priority": 0,
  },
  {
    "diagnosis": "bleeding",
    "name": "Jane",
    "priority": 2,
  },
  {
    "diagnosis": "flu",
    "name": "May",
    "priority": 4,
  },
  {
    "diagnosis": "injury",
    "name": "Simon",
    "priority": 3,
  },
  {
    "diagnosis": "flu",
    "name": "Nick",
    "priority": 4,
  },
  {
    "diagnosis": "sniffing",
    "name": "Aaron",
    "priority": 6,
  },
  {
    "diagnosis": "high temperature",
    "name": "Jack",
    "priority": 8,
  },
  {
    "diagnosis": "flu",
    "name": "Dread",
    "priority": 4,
  },
]
`;
